applepay/utils.py

```diff
@@ -66,3 +85,295 @@ def signing_time_is_valid(signature, current_time, threshold):
+    Returns:
+        callable: The callable hash function of the provided
+            algorithm updated with the data to be hashed
+    """
+    hashfunc = hashlib.new(name)
+    hashfunc.update(data)
+    return hashfunc
+
+
+def verify_signature(token):
+    """Verify the signature within an apple pay token according to
+    the #ApplePaySpec documentation."""
+
+    # We only bother to validate EC_v1 at this time.  RSA is only used
+    # for transactions from China and is not supported at this time.
+    if token['version'] != 'EC_v1':
+        logger.warning("Unsupported version {}".format(token['version']))
+        return False
+
+    # extract and decode the signature object
+    # into a CMS object
+    signature = token['signature']
+    signature_data = base64.b64decode(signature)
+    content_info = cms.ContentInfo.load(signature_data)
+    signed_data = content_info['content']
+    certificates = signed_data['certificates']
+
+    # there should only be 2 certificates present
+    if len(certificates) != 2:
+        logger.warning("Expected 2 certificates, found {}".format(len(certificates)))
+        return False
+
+    leaf_cert, intermediate_cert = get_leaf_and_intermediate_certs(certificates)
+
+    if not leaf_cert:
+        logger.warning("Leaf Certificate OID not found")
+        return False
+
+    if not intermediate_cert:
+        logger.warning("Intermediate certificate OID not found")
+        return False
+
+    root_der = open(payment.ROOT_CA_FILE, 'r').read()
+
+    # pass the der-encoded representation of the certs
+    if not valid_chain_of_trust(root_der, intermediate_cert.dump(), leaf_cert.dump()):
+        return False
+
+    # Build the signer information from the signer that matches the leaf cert
+    signer = get_first_from_iterable(
+        filter_func=lambda signer: signer['sid'].chosen['serial_number'].native == leaf_cert.serial_number,
+        iterable=signed_data['signer_infos']
+    )
+
+    if not signer:
+        logger.warning("No signature found for the leaf cert.")
+        return False
+
+    # Use the signed attrs to verify the data is from who it says its from
+    signed_attrs = signer['signed_attrs']
+
+    message_digest_attr = get_first_from_iterable(
+        filter_func=lambda signed_attr: signed_attr['type'].dotted == payment.OID_MESSAGE_DIGEST,
+        iterable=signed_attrs
+    )
+
+    if not message_digest_attr:
+        logger.warning("No message digest found for the leaf cert.")
+        return False
+
+    message_digest = message_digest_attr['values'][0].native
+
+    data = get_payment_data(token)
+
+    # build the hashfunc from the leaf_cert's defined algorithm
+    hashfunc = partial(get_hashfunc_by_name, leaf_cert.hash_algo)
+
+    hashed_data = hashfunc(data).digest()
+
+    if hashed_data != message_digest:
+        logger.warning("Message digest does not match provided data.")
+        return False
+
+    signed_data = get_signed_data(signed_attrs)
```

Comment:
I update this function and variable name to be more closely tied to what its doing; returning the ber-encoded signed attributes.